tools/cmake/cbmc_recipes.cmake

```diff
@@ -0,0 +1,323 @@
+# ``````````````````````````````````````````````````````````````````````````````
+#               Common recipes for building and running CBMC proofs
+#               ===================================================
+#
+# This recipes file should be included at the bottom of every CMakeLists.txt
+# that is used for defining a CBMC proof. The CMakeLists.txt is responsible for
+# setting the options for that particular proof, while this recipes file uses
+# those options to actually generate the builds for that proof. This file uses
+# the variable definitions in tools/cmake/toolchains/cmake.cbmc, which can be
+# changed by the CMakeLists.txt of each proof.
+#
+# This file builds the object files for each proof (consisting of the proof
+# harness plus the code that is being proven); does some processing on those
+# object files (removing unneeded functions, adding in function stubs, etc); and
+# then generates 'tests' over those object files that actually run CBMC and
+# generate a HTML proof report.
+# ______________________________________________________________________________
+
+
+# ______________________________________________________________________________
+#     Sanity checks: ensure that the CMakeLists.txt for a single proof is
+#     correctly written
+# ``````````````````````````````````````````````````````````````````````````````
+
+if(NOT DEFINED cbmc_proof_name)
+    message(FATAL_ERROR "\
+The CMakeLists.txt for the CBMC proof in directory
+${CMAKE_CURRENT_SOURCE_DIR}
+should set the variable 'cbmc_proof_name' to the name of the proof.")
+endif()
+
+if(${cbmc_proof_name} IN_LIST cbmc_proof_names)
+    message(FATAL_ERROR "\
+The CMakeLists.txt for the CBMC proof in directory
+${CMAKE_CURRENT_SOURCE_DIR}
+sets the proof name to '${cbmc_proof_name}', but that name is already used by
+another proof. Ensure that you set the variable 'cbmc_proof_name' to a unique
+value for each proof.")
+endif()
+
+list(LENGTH cbmc_nondet_static_exclude cnse_length)
+if(${cnse_length} AND ${cbmc_nondet_static})
+    message(FATAL_ERROR "\
+The CMakeLists.txt for the CBMC proof in directory
+${CMAKE_CURRENT_SOURCE_DIR}
+sets both `cbmc_nondet_static` to true, and `cbmc_nondet_static_exclude` to a
+nonempty list. You should only do one or the other. Value of
+cbmc_nondet_static_exclude: [${cbmc_nondet_static_exclude}].")
+endif()
+
+set(cbmc_proof_names "${cbmc_proof_name};${cbmc_proof_names}" CACHE INTERNAL "")
```

Comment:
Should this be list append?

Reply:
It seems that `list(APPEND` doesn't work on cache variables